### PR TITLE
is_undef for checking for undef without a warning

### PR DESCRIPTION
**Pull request by MichaelPFrey**
this goes back to #2524 and #2468 

the ~~current~~ first implementation ~~is~~ was a hack, but it already shows how it could work and where the tweaks have to go.

The todo list beeing:
- [x] documentation
- [x] syntax highlighting
- [x]  more test cases
   - [x]  multiple parameters?
      - [x] isundef(1,2)
   - [x] "calculation"/more complex expression
      - [x] isundef(log(1,2,3))
      - [x] isundef(len("test","upps"))
      - [x] isundef(isundef())
- [x]  `src/evalcontext.h` // `const AssignmentList &eval_arguments;` shall remain private
- [x]  isundef -> is_undef
with now just the cheat sheet beeing open.

### CODE REVIEW

**Comment by t-paul** [comment]
Maybe `is_undef()` matching the #1584 request for other check functions?

**Comment by MichaelPFrey** [comment]
Documentation: (tobe copied to [wikibooks](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/WIP))

is_undef accepts one parameter. If the parameter is undef, this function returns true. If the parameter is not undef, it returns false. When checking a variable (like `is_undef(a)`), it does the variable lookup silently, meaning that is_undef(a) does not cause `WARNING: Ignoring unknown variable 'a'. `

The alternative is code like this:
```
if(a==undef){
    //code goes here
}
```
or
```
b = (a==undef) ? true : false;
```
will cause `WARNING: Ignoring unknown variable 'a'. `.

is_undef also works for special variables, allowing for things like this:
```
exploded = is_undef($exploded) ? 0 : $exploded; // 1 for exploded view
```
----
For older openscad version, is_undef can be emulated with
`function is_undef ( a ) = (undef == a) ;`
which off-course causes warning(s), but requires no changes to code relaying on is_undef().

File: src/func.cc [hunk 1]
```diff
@@ -924,6 +924,22 @@ ValuePtr builtin_cross(const Context *, const EvalContext *evalctx)
 	return ValuePtr(result);
 }
 
+ValuePtr builtin_is_undef(const Context *, const EvalContext *evalctx)
+{
+	if (evalctx->numArgs() == 1) {
+		const auto &arg =evalctx->getArgs()[0];
+		ValuePtr v;
+		if(const Lookup* lookup = dynamic_cast<const Lookup*> (arg.expr.get())){
```
Comment (kintel) [inline src/func.cc]: Tip: You can use `dynamic_pointer_cast` instead of unpacking the shared_ptr

**Comment by kintel** [comment]
Looks good. Unless it has already been done, we also need to make a reminder to update cheatsheet and docs before the next release.

File: src/func.cc (hunk 1 above)
Comment (MichaelPFrey) [reply src/func.cc]: That is much nicer. And there is no need to repeat the types, so auto simplifies it further.

**Comment by MichaelPFrey** [comment]
Documentation will follow. I prefer to have a consent about the idea and have the implementation merged, before adding a feature to the documentation. Once its documented, people may start anticipating.

**Pull request merged**